### PR TITLE
Observe shadow root on connected callback

### PR DESCRIPTION
**Pull request by williammartin**
## Description

While poking around and learning about catalyst I tried to use [dynamic rendering](https://catalyst.rocks/guide/rendering/) on the shadow DOM for both a `parent` and a `child` custom web component. I was able to see the parent attach the child but the contents of the child were not attached to the DOM.

### Why tho?

The existing `lazyDefine` creates a `MutationObserver` that listens to the entire `document`, but shadow roots live as a separate isolated node and so when the shadow root of the `parent` contained `child`, the import callback held in `dynamicElements` was not being called.

### What's the fix?

We exported a new `observe` function which will observe for changes on the provided `ElementLike`. Then in the controller `connectedCallback`, if there is a `shadowRoot`, we'll observe it.

CC @keithamus who I paired on this with

### CODE REVIEW

File: docs/_guide/rendering-2.md [hunk 1]
```diff
@@ -89,7 +89,7 @@ class HelloWorldElement extends HTMLElement {
   }
 
   attributeChangedCallback() {
-    render(() => html`
```
Comment (williammartin) [inline docs/_guide/rendering-2.md:92]: These are unrelated to the main change, just type errors.

**Comment by primer-css** [comment]
:wave: Hello and thanks for pinging us! This issue or PR has been added to our inbox and a Design Infrastructure first responder will review it soon.
- :art: **If this is a PR that includes a visual change**, please make sure to add screenshots in the description or deploy this code to a lab machine with instructions for how to test.
- :fast_forward: **If this is a PR that includes changes to an interaction**, please include a [video recording](https://support.apple.com/en-us/HT208721) in the description.
- :warning: **If this is urgent**, please visit us in [#primer](https://github.slack.com/archives/CSGAVNZ19) on Slack and tag the first responders listed in the channel topic.

**Comment by keithamus** [comment]
Great work @williammartin! LGTM :wink:

**Review by keithamus — APPROVED** [review]
